### PR TITLE
Separate the Gradient Accumulation Scheduler from Trainer

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Removes support for passing a scheduling dictionary to `Trainer(accumulate_grad_batches=...)`. The same functionality can still be achieved by simply passing the callback in. This simplifies the Trainer and the overall validation logic. Furthermore, it makes the callback self-contained and leaves open the possibility to move it out of the lightning package (not decided). 

Before:
```py
trainer = Trainer(accumulate_grad_batches={"1": 5, "10": 3})
```

Now:
```py
from lightning.pytorch.callbacks import GradientAccumulationScheduler
trainer = Trainer(callbacks=GradientAccumulationScheduler({"1": 5, "10": 3}))
```

This refactor and breaking change to the Trainer API is part of the Trainer 2.0 simplification effort. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @justusschock @awaelchli @borda

### CODE REVIEW

File: src/lightning/pytorch/callbacks/gradient_accumulation_scheduler.py [hunk 1]
```diff
@@ -99,5 +98,44 @@ def get_accumulate_grad_batches(self, epoch: int) -> int:
                 break
         return accumulate_grad_batches
 
+    def on_train_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
```
Comment (awaelchli) [inline src/lightning/pytorch/callbacks/gradient_accumulation_scheduler.py:101]: The validation of settings now happens here instead of Trainer's config validator so that the entire feature is self-contained in the callback.

File: tests/tests_pytorch/accelerators/test_ipu.py [hunk 2]
```diff
@@ -319,16 +316,6 @@ def on_predict_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, da
     trainer.predict(model, model.test_dataloader())
 
 
-@RunIf(ipu=True)
```
Comment (awaelchli) [inline tests/tests_pytorch/accelerators/test_ipu.py:322]: The tests for unsupported accelerators and strategies has been moved to the callback tests below. No need to run on the actual accelerator or launch expensive processes to test this.

File: tests/tests_pytorch/callbacks/test_gradient_accumulation_scheduler.py [hunk 3]
```diff
@@ -57,33 +57,15 @@ def test_trainer_accumulate_grad_batches_dict_zero_grad(tmpdir, accumulate_grad_
...
         )
-        assert trainer.accumulate_grad_batches == accumulate_grad_batches.get(0, 1)
+        assert trainer.accumulate_grad_batches == 1  # initial value of Trainer argument
         trainer.fit(model)
 
         assert sum(isinstance(cb, GradientAccumulationScheduler) for cb in trainer.callbacks) == 1
         assert sgd_zero_grad.call_count == expected_call_count
 
 
-def test_trainer_accumulate_grad_batches_with_callback(tmpdir):
```
Comment (awaelchli) [inline tests/tests_pytorch/callbacks/test_gradient_accumulation_scheduler.py:69]: the test above already includes this case now

File: src/lightning/pytorch/trainer/trainer.py [hunk 4]
```diff
@@ -1832,14 +1833,6 @@ def configure_optimizers(self):
                 return [optimizer], [scheduler]
 
         """
-        accumulation_scheduler = self.accumulation_scheduler
-
-        if accumulation_scheduler.epochs != [0]:
-            raise MisconfigurationException(
-                "Estimated stepping batches cannot be computed with different"
-                " `accumulate_grad_batches` at different epochs."
-            )
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning/pytorch/trainer/trainer.py:1841]: I think this is going too far. In general, any callback can just change the trainer state arbitrarily.The task that `Trainer.estimated_stepping_batches` has is ill posed. It cannot possibly produce an accurate answer if we allow callbacks to arbitrarily change how long an epoch is at any time during execution.

File: src/lightning/pytorch/trainer/connectors/callback_connector.py [hunk 5]
```diff
@@ -76,49 +74,13 @@ def on_trainer_init(
...
-            cb for cb in self.trainer.callbacks if isinstance(cb, GradientAccumulationScheduler)
-        ]
-
-        if grad_accum_callbacks:
-            if accumulate_grad_batches is not None:
-                raise MisconfigurationException(
-                    "You have set both `accumulate_grad_batches` and passed an instance of "
-                    "`GradientAccumulationScheduler` inside callbacks. Either remove `accumulate_grad_batches` "
-                    "from trainer or remove `GradientAccumulationScheduler` from callbacks list."
-                )
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning/pytorch/trainer/connectors/callback_connector.py:102]: Ok, but I prefer to keep the default 1 instead of None. Easier to read and document.

File: src/lightning/pytorch/trainer/trainer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply src/lightning/pytorch/trainer/trainer.py:1841]: That's true, but adding this check doesnt cost us anything and could avoid a costly unintended bug for whoever uses both feature together. Sure it's only valid for this callback.

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]
